syntax = "proto3";

service DBService {
    rpc Session (stream SessionRequest) returns (stream SessionResponse);

    rpc GetStatus (GetStatusRequest) returns (GetStatusReply) {}

    // Database
    rpc ListAvailableDB (ListAvailableDBRequest) returns (ListAvailableDBReply) {}
    rpc ListLoadedDB (ListLoadedDBRequest) returns (ListLoadedDBReply) {}
    rpc LoadDB (LoadDBRequest) returns (LoadDBReply) {}
    rpc UnloadDB (UnloadDBRequest) returns (UnloadDBReply) {}
    rpc CreateDB (CreateDBRequest) returns (CreateDBReply) {}
    rpc DumpDB (DumpDBRequest) returns (DumpDBReply) {}

    // NodeTypes
    rpc ListNodeTypes (ListNodeTypesRequest) returns (ListNodeTypesReply) {}
    rpc ListNodeTypesByID (ListNodeTypesByIDRequest) returns (ListNodeTypesByIDReply) {}
    rpc CreateNodeType (CreateNodeTypeRequest) returns (CreateNodeTypeReply) {}

    // EdgeTypes
    rpc ListEdgeTypes (ListEdgeTypesRequest) returns (ListEdgeTypesReply) {}
    rpc ListEdgeTypesByID (ListEdgeTypesByIDRequest) returns (ListEdgeTypesByIDReply) {}
    rpc CreateEdgeType (CreateEdgeTypeRequest) returns (CreateEdgeTypeReply) {}

    // PropertyTypes
    rpc AddPropertyType (AddPropertyTypeRequest) returns (AddPropertyTypeReply) {}
    rpc ListPropertyTypes (ListPropertyTypesRequest) returns (ListPropertyTypesReply) {}

    // Networks
    rpc ListNetworks (ListNetworksRequest) returns (ListNetworksReply) {}
    rpc CreateNetwork (CreateNetworkRequest) returns (CreateNetworkReply) {}

    // Nodes
    rpc ListNodes (ListNodesRequest) returns (ListNodesReply) {}
    rpc ListNodesByID (ListNodesByIDRequest) returns (ListNodesByIDReply) {}
    rpc ListNodesFromNetwork (ListNodesFromNetworkRequest) returns (ListNodesFromNetworkReply) {}
    rpc ListNodesByIDFromNetwork (ListNodesByIDFromNetworkRequest) returns (ListNodesByIDFromNetworkReply) {}
    rpc CreateNode (CreateNodeRequest) returns (CreateNodeReply) {}

    rpc ListEntityProperties (ListEntityPropertiesRequest) returns (ListEntityPropertiesReply) {}
    rpc AddEntityProperty (AddEntityPropertyRequest) returns (AddEntityPropertyReply) {}

    // Edges
    rpc ListEdges (ListEdgesRequest) returns (ListEdgesReply) {}
    rpc ListEdgesByID (ListEdgesByIDRequest) returns (ListEdgesByIDReply) {}
    rpc ListEdgesFromNetwork (ListEdgesFromNetworkRequest) returns (ListEdgesFromNetworkReply) {}
    rpc ListEdgesByIDFromNetwork (ListEdgesByIDFromNetworkRequest) returns (ListEdgesByIDFromNetworkReply) {}
    rpc CreateEdge (CreateEdgeRequest) returns (CreateEdgeReply) {}
}

// Structs
message Database {
    string name = 1;
    uint64 id = 2;
}

message Network {
    uint64 db_id = 1;
    uint64 id = 2;
    string name = 3;
}

message EdgeDefinition {
    uint64 edge_id = 1;
    uint64 node_id = 2;
}

message Node {
    uint64 db_id = 1;
    uint64 net_id = 2;
    uint64 id = 3;
    string name = 4;
    uint64 node_type_id = 5;
    uint64 in_edge_count = 6;
    uint64 out_edge_count = 7;
    repeated EdgeDefinition in_edges = 8;
    repeated EdgeDefinition out_edges = 9;
    repeated Property properties = 10;
}

message Edge {
    uint64 db_id = 1;
    uint64 id = 2;
    uint64 edge_type_id = 3;
    uint64 source_id = 4;
    uint64 target_id = 5;
    repeated Property properties = 6;
}

message NodeType {
    uint64 db_id = 1;
    uint64 id = 2;
    string name = 3;
    repeated uint64 in_edge_ids = 4;
    repeated uint64 out_edge_ids = 5;
}

message EdgeType {
    uint64 db_id = 1;
    uint64 id = 2;
    string name = 3;
    repeated uint64 source_ids = 4;
    repeated uint64 target_ids = 5;
}

enum ValueType {
    INVALID = 0;
    INT = 1;
    UNSIGNED = 2;
    BOOL = 3;
    DECIMAL = 4;
    STRING_REF = 5;
    STRING = 6;
}

enum EntityType {
    NODE = 0;
    EDGE = 1;
}

message Property {
    string property_type_name = 1;
    ValueType value_type = 2;
    oneof value {
        string string = 3;
        int64 int64 = 4;
        uint64 uint64 = 5;
        bool boolean = 6;
        double decimal = 7;
    };
}

message PropertyType {
    string name = 1;
    ValueType value_type = 2;
}

message FilterID {
    repeated uint64 ids = 1;
}

message FilterType {
    oneof value {
        uint64 node_type_id = 1;
        uint64 edge_type_id = 2;
    };
}

message FilterProperty {
    Property property = 1;
}

// GetStatus
message GetStatusRequest {}
message GetStatusReply {}

// ListAvailableDB
message ListAvailableDBRequest {}
message ListAvailableDBReply {
    repeated string db_names = 1;
}

// ListLoadedDB
message ListLoadedDBRequest {}
message ListLoadedDBReply {
    map<string, Database> dbs = 1;
}

// LoadDB
message LoadDBRequest {
    string db_name = 1;
}
message LoadDBReply {
    Database db = 1;
}

// UnloadDB
message UnloadDBRequest {
    string db_name = 1;
}
message UnloadDBReply {
}

// CreateDB
message CreateDBRequest {
    string db_name = 1;
}
message CreateDBReply {
    Database db = 1;
}

// DumpDB
message DumpDBRequest {
    uint64 db_id = 1;
}
message DumpDBReply {}

// ListNetworks
message ListNetworksRequest {
    uint64 db_id = 1;
}
message ListNetworksReply {
    repeated Network networks = 1;
}

// ListNodes
message ListNodesRequest {
    uint64 db_id = 1;
    optional FilterType filter_type = 2;
    optional FilterID filter_id = 3;
    optional FilterProperty filter_property = 4;
    bool yield_edges = 5;
}
message ListNodesReply {
    repeated Node nodes = 1;
}

// ListNodesByID
message ListNodesByIDRequest {
    uint64 db_id = 1;
    repeated uint64 node_ids = 2;
    bool yield_edges = 3;
    uint64 max_edge_count = 4;
    repeated string edge_type_filter_out = 5;
    repeated string node_type_filter_out = 6;
    repeated Property node_property_filter_out = 7;
    repeated Property node_property_filter_in = 8;
}
message ListNodesByIDReply {
    repeated Node nodes = 1;
}

// ListEdges
message ListEdgesRequest {
    uint64 db_id = 1;
}
message ListEdgesReply {
    repeated Edge edges = 1;
}

// ListEdgesByID
message ListEdgesByIDRequest {
    uint64 db_id = 1;
    repeated uint64 edge_ids = 2;
}
message ListEdgesByIDReply {
    repeated Edge edges = 1;
}

// ListNodesFromNetwork
message ListNodesFromNetworkRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
}
message ListNodesFromNetworkReply {
    repeated Node nodes = 1;
}

// ListNodesByIDFromNetwork
message ListNodesByIDFromNetworkRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
    repeated uint64 node_ids = 3;
}
message ListNodesByIDFromNetworkReply {
    repeated Node nodes = 1;
}

// ListEdgesFromNetwork
message ListEdgesFromNetworkRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
}
message ListEdgesFromNetworkReply {
    repeated Edge edges = 1;
}

// ListEdgesByIDFromNetwork
message ListEdgesByIDFromNetworkRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
    repeated uint64 edge_ids = 3;
}
message ListEdgesByIDFromNetworkReply {
    repeated Edge edges = 1;
}

// ListNodeTypes
message ListNodeTypesRequest {
    uint64 db_id = 1;
}
message ListNodeTypesReply {
    repeated NodeType node_types = 1;
}

// ListNodeTypesByID
message ListNodeTypesByIDRequest {
    uint64 db_id = 1;
    repeated uint64 node_type_ids = 2;
}
message ListNodeTypesByIDReply {
    repeated NodeType node_types = 1;
}

// ListEdgeTypes
message ListEdgeTypesRequest {
    uint64 db_id = 1;
}
message ListEdgeTypesReply {
    repeated EdgeType edge_types = 1;
}

// ListEdgeTypesByID
message ListEdgeTypesByIDRequest {
    uint64 db_id = 1;
    repeated uint64 edge_type_ids = 2;
}
message ListEdgeTypesByIDReply {
    repeated EdgeType edge_types = 1;
}

// CreateNetwork
message CreateNetworkRequest {
    uint64 db_id = 1;
    string name = 2;
}
message CreateNetworkReply {
    Network network = 1;
}

// CreateNodeType
message CreateNodeTypeRequest {
    uint64 db_id = 1;
    string name = 2;
}
message CreateNodeTypeReply {
    NodeType node_type = 1;
}

// CreateEdgeType
message CreateEdgeTypeRequest {
    uint64 db_id = 1;
    string name = 2;
    repeated uint64 source_ids = 3;
    repeated uint64 target_ids = 4;
}
message CreateEdgeTypeReply {
    EdgeType edge_type = 1;
}

// CreateNode
message CreateNodeRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
    string name = 3;
    uint64 node_type_id = 4;
}
message CreateNodeReply {
    Node node = 1;
}

// CreateEdge
message CreateEdgeRequest {
    uint64 db_id = 1;
    uint64 net_id = 2;
    uint64 edge_type_id = 3;
    uint64 source_id = 4;
    uint64 target_id = 5;
}
message CreateEdgeReply {
    Edge edge = 1;
}

// AddEntityProperty
message AddEntityPropertyRequest {
    uint64 db_id = 1;
    uint64 entity_id = 2;
    EntityType entity_type = 3;
    Property property = 5;
}
message AddEntityPropertyReply {
    Property property = 1;
}

// ListEntityProperties
message ListEntityPropertiesRequest {
    uint64 db_id = 1;
    EntityType entity_type = 2;
    uint64 entity_id = 3;
}
message ListEntityPropertiesReply {
    repeated Property properties = 1;
}

// ListEdgeProperties
message ListEdgePropertiesRequest {
    uint64 db_id = 1;
    uint64 edge_id = 2;
}
message ListEdgePropertiesReply {
    repeated Property properties = 1;
}

// AddPropertyType
message AddPropertyTypeRequest {
    uint64 db_id = 1;
    uint64 entity_type_id = 2;
    EntityType entity_type = 3;
    PropertyType property_type = 4;
}
message AddPropertyTypeReply {
    PropertyType property_type = 1;
}

// ListPropertyTypes
message ListPropertyTypesRequest {
    uint64 db_id = 1;
    uint64 entity_type_id = 2;
    EntityType entity_type = 3;
}
message ListPropertyTypesReply {
    repeated PropertyType property_types = 1;
}

message SessionRequest {
    oneof req {
        OpenSession open_req = 1;
        ExecuteQuery query_req = 2;
        PullRequest pull_req = 3;
    }
}

message SessionResponse {
    oneof res {
        Success success_res = 1;
        Failure failure_res = 2; 
        PullResponse pull_res = 3;
    }
}

message OpenSession {
    string token = 1;
}

message ExecuteQuery {
    string query_str = 1;
}

message PullRequest {
    uint64 query_id = 1;
}

message Success {
    uint64 query_id = 1;
}

message Failure {
    string msg = 1;
}

message PullResponse {
    repeated Record records = 1;
}

message Record {
    repeated Cell cells = 1;
}

message Cell {
    oneof value {
        string string = 3;
        int64 int64 = 4;
        uint64 uint64 = 5;
        bool boolean = 6;
        double decimal = 7;
    }
}


